File: Psiphon/ViewControllers/OnboardingViewController.m
/*
 * Copyright (c) 2018, Psiphon Inc.
 * All rights reserved.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#import <ReactiveObjC/RACSignal.h>
#import "OnboardingViewController.h"
#import "OnboardingView.h"
#import "UIColor+Additions.h"
#import "SkyButton.h"
#import "RingSkyButton.h"
#import "UIFont+Additions.h"
#import "RoyalSkyButton.h"
#import "VPNManager.h"
#import "RACCompoundDisposable.h"
#import "AlertDialogs.h"
#import "UIView+Additions.h"
#import "OnboardingScrollableView.h"
#import "Strings.h"
#import "ContainerDB.h"
#import "LanguageSelectionViewController.h"
#import "AppDelegate.h"
#import "CloudsView.h"
#import "Logging.h"

const int NumPages = 4;

@interface OnboardingViewController ()

@property (nonatomic) RACCompoundDisposable *compoundDisposable;

@end

@implementation OnboardingViewController {
    CloudsView *cloudsView;
    UIProgressView *progressView;
    UIButton *nextPageButton;

    UIView *currentOnboardingView;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        _compoundDisposable = [RACCompoundDisposable compoundDisposable];
    }
    return self;
}

- (void)dealloc {
    [self.compoundDisposable dispose];
}

- (UIStatusBarStyle)preferredStatusBarStyle {
    return UIStatusBarStyleLightContent;
}

- (void)viewDidLoad {
    [super viewDidLoad];

    [self setNeedsStatusBarAppearanceUpdate];

    self.view.backgroundColor = UIColor.darkBlueColor;

    // Clouds view
    {
        cloudsView = [[CloudsView alloc] initForAutoLayout];
        [self.view addSubview:cloudsView];
        [NSLayoutConstraint activateConstraints:@[
          [cloudsView.topAnchor constraintEqualToAnchor:self.view.safeTopAnchor],
          [cloudsView.leadingAnchor constraintEqualToAnchor:self.view.safeLeadingAnchor],
          [cloudsView.trailingAnchor constraintEqualToAnchor:self.view.safeTrailingAnchor],
          [cloudsView.bottomAnchor constraintEqualToAnchor:self.view.safeBottomAnchor]
        ]];
    }

    // Progress view
    {
        progressView = [[UIProgressView alloc] initWithProgressViewStyle:UIProgressViewStyleBar];
        [self.view addSubview:progressView];
        progressView.progressTintColor = UIColor.lightishBlue;

        progressView.translatesAutoresizingMaskIntoConstraints = FALSE;
        [NSLayoutConstraint activateConstraints:@[
          [progressView.bottomAnchor constraintEqualToAnchor:self.view.safeBottomAnchor],
          [progressView.leadingAnchor constraintEqualToAnchor:self.view.safeLeadingAnchor],
          [progressView.trailingAnchor constraintEqualToAnchor:self.view.safeTrailingAnchor],
          [progressView.heightAnchor constraintEqualToConstant:6.f]
        ]];
    }

    // Next button
    {
        nextPageButton = [UIButton buttonWithType:UIButtonTypeSystem];
        CGFloat p = 20.f; // Margin around the next button.
        nextPageButton.contentEdgeInsets = UIEdgeInsetsMake(p, 2.f * p, p, p);

        [nextPageButton setTitle:[Strings nextPageButtonTitle] forState:UIControlStateNormal];
        [nextPageButton setTitleColor:UIColor.whiteColor forState:UIControlStateNormal];
        nextPageButton.titleLabel.font = [UIFont avenirNextDemiBold:16.f];
        [nextPageButton addTarget:self
                 action:@selector(gotoNextPage)
       forControlEvents:UIControlEventTouchUpInside];
        [self.view addSubview:nextPageButton];

        nextPageButton.translatesAutoresizingMaskIntoConstraints = FALSE;
        [NSLayoutConstraint activateConstraints:@[
          [nextPageButton.bottomAnchor constraintEqualToAnchor:progressView.topAnchor],
          [nextPageButton.trailingAnchor constraintEqualToAnchor:self.view.safeTrailingAnchor
                                                        constant:-20.f],
        ]];
    }

    UIView *firstPage = [self createOnboardingViewForPage:1];
    [self setCurrentOnboardingPage:firstPage];
}

#pragma mark - Page update methods

- (void)setCurrentOnboardingPage:(UIView *_Nonnull)onboardingPage {
    progressView.progress = (CGFloat) (onboardingPage.tag) / (CGFloat) NumPages;

    if (onboardingPage.tag == 1 ||
        onboardingPage.tag == NumPages - 1) {
        nextPageButton.hidden = FALSE;
    } else {
        nextPageButton.hidden = TRUE;
    }

    if (currentOnboardingView) {
        [currentOnboardingView removeFromSuperview];
    }

    currentOnboardingView = onboardingPage;
    [self.view addSubview:currentOnboardingView];
    [self applyOnboardingViewConstraintsToView:currentOnboardingView
                  anchorBottomToNextPageButton:!nextPageButton.hidden];
}

- (UIView *_Nullable)createOnboardingViewForPage:(NSInteger)page {
    UIView *v;

    switch (page) {
        case 1: {
            // Language selection page
            RingSkyButton *selectLangButton = [[RingSkyButton alloc] initForAutoLayout];
            selectLangButton.includeChevron = TRUE;
            [selectLangButton setTitle:[Strings onboardingSelectLanguageButtonTitle]];
            [selectLangButton addTarget:self
                                 action:@selector(onLanguageSelectionButton)
                       forControlEvents:UIControlEventTouchUpInside];

            v = [[OnboardingView alloc]
              initWithImage:[UIImage imageNamed:@"OnboardingStairs"]
                  withTitle:[Strings onboardingBeyondBordersHeaderText]
                   withBody:[Strings onboardingBeyondBordersBodyText]
          withAccessoryView:selectLangButton];
            break;
        }
        case 2: {
            // Privacy Policy page
            RoyalSkyButton *acceptButton = [[RoyalSkyButton alloc] initForAutoLayout];
            [acceptButton setTitle:[Strings acceptButtonTitle]];
            [acceptButton addTarget:self
                             action:@selector(onPrivacyPolicyAccepted)
                   forControlEvents:UIControlEventTouchUpInside];
            acceptButton.shadow = TRUE;

            RingSkyButton *declineButton = [[RingSkyButton alloc] initForAutoLayout];
            [declineButton setTitle:[Strings declineButtonTitle]];
            [declineButton addTarget:self
                              action:@selector(onPrivacyPolicyDeclined)
                    forControlEvents:UIControlEventTouchUpInside];

            UIStackView *buttonsView = [[UIStackView alloc]
              initWithArrangedSubviews:@[declineButton, acceptButton]];
            buttonsView.spacing = 20.f;
            buttonsView.distribution = UIStackViewDistributionFillEqually;

            v = [[OnboardingScrollableView alloc]
              initWithImage:[UIImage imageNamed:@"OnboardingPrivacyPolicy"]
                  withTitle:[Strings privacyPolicyTitle]
                   withBody:[Strings privacyPolicyHTMLText]
          withAccessoryView:buttonsView];

            break;
        }
        case 3: {
            // "Getting Started" page
            v = [[OnboardingView alloc]
              initWithImage:[UIImage imageNamed:@"OnboardingPermission"]
                  withTitle:[Strings onboardingGettingStartedHeaderText]
                   withBody:[Strings onboardingGettingStartedBodyText]
          withAccessoryView:nil];
            break;
        }
        //
        case 4: {
            v = [self createInstallGuideVPNOnboardingView];
            break;
        }

        default:
            return nil;
    }

    v.tag = page;
    return v;
}

- (void)gotoNextPage {
    OnboardingViewController *__weak weakSelf = self;

    // Present the next onboarding page. If none is available notify the delegate
    // that the onboarding has finished.
    UIView *_Nullable newPage = [self createOnboardingViewForPage:currentOnboardingView.tag + 1];
    if (newPage) {
        [self setCurrentOnboardingPage:newPage];
    } else {
        if (self.delegate) {
            [self.delegate onboardingFinished:self];
        }
        return;
    }

    // Prompt user for VPN configuration permission on page 4.
    if (newPage.tag == 4) {

        __block RACDisposable *disposable = [[[VPNManager sharedInstance] reinstallVPNConfiguration]
          subscribeNext:^(RACUnit *x) {
              // Go to next onboarding page.
              [weakSelf gotoNextPage];
          }
          error:^(NSError *error) {
              // Go to previous page.
              UIView *prvPage = [weakSelf createOnboardingViewForPage:3];
              [weakSelf setCurrentOnboardingPage:prvPage];

              // If the error was due to user denying permission to install VPN configuration,
              // shows the `vpnPermissionDeniedAlert` instead of the generic operation failed alert.
              if ([error.domain isEqualToString:NEVPNErrorDomain] &&
                  error.code == NEVPNErrorConfigurationReadWriteFailed &&
                  [error.localizedDescription isEqualToString:@"permission denied"] ) {

                  // Present the VPN permission denied alert.
                  UIAlertController *alert = [AlertDialogs vpnPermissionDeniedAlert];
                  [weakSelf presentViewController:alert animated:TRUE completion:nil];

              } else {
                  UIAlertController *alert = [AlertDialogs genericOperationFailedTryAgain];
                  [weakSelf presentViewController:alert animated:TRUE completion:nil];
              }
              [weakSelf.compoundDisposable removeDisposable:disposable];
          }
          completed:^{
              [weakSelf.compoundDisposable removeDisposable:disposable];
          }];

        [self.compoundDisposable addDisposable:disposable];
    }
}

#pragma mark - UI callbacks

- (void)onLanguageSelectionButton {

    LanguageSelectionViewController *vc = [[LanguageSelectionViewController alloc]
      initWithSupportedLanguages];

    vc.selectionHandler = ^(NSUInteger selectedIndex, id selectedItem,
      PickerViewController *viewController) {
        [viewController dismissViewControllerAnimated:TRUE completion:nil];
        // Reload the onboarding to reflect the newly selected language.
        [[AppDelegate sharedAppDelegate] reloadOnboardingViewController];
    };

    UINavigationController *nav = [[UINavigationController alloc] initWithRootViewController:vc];

    [self presentViewController:nav animated:TRUE completion:nil];
}

- (void)onPrivacyPolicyAccepted {
    // Stores the privacy policy date that the user accepted.
    ContainerDB *containerDB = [[ContainerDB alloc] init];
    [containerDB setAcceptedPrivacyPolicyUnixTime:[containerDB privacyPolicyLastUpdateTime]];
    // Go to next page;
    [self gotoNextPage];
}

- (void)onPrivacyPolicyDeclined {
    // Only alert the user that they need to accept the privacy policy.
    UIAlertController *alert = [AlertDialogs privacyPolicyDeclinedAlert];
    [self presentViewController:alert animated:TRUE completion:nil];
}

#pragma mark - View helper methods

// Creates view with an arrow pointing to "Allow" button of permission dialog.
// The X and Y center of the returned view is expected to match the X and Y center of the screen.
- (UIView *)createInstallGuideVPNOnboardingView {
    UIView *view = [[UIView alloc] init];

    UIImage *arrowImage = [UIImage imageNamed:@"PermissionArrow"];
    UIImageView *arrowView = [[UIImageView alloc] initWithImage:arrowImage];
    [view addSubview:arrowView];

    UILabel *label = [[UILabel alloc] initWithFrame:CGRectZero];
    label.text = [Strings vpnInstallGuideText];
    [view addSubview:label];

    label.adjustsFontSizeToFitWidth = TRUE;
    label.minimumScaleFactor = 0.8;
    label.font = [UIFont avenirNextMedium:16.f];
    label.textColor = UIColor.whiteColor;
    label.numberOfLines = 0;
    label.lineBreakMode = NSLineBreakByWordWrapping;
    label.textAlignment = NSTextAlignmentCenter;

    CGFloat aspectRatio = (arrowImage.size.width / arrowImage.size.height);
    arrowView.translatesAutoresizingMaskIntoConstraints = FALSE;
    [NSLayoutConstraint activateConstraints:@[
      [arrowView.heightAnchor constraintEqualToConstant:84.f],
      [arrowView.widthAnchor constraintEqualToAnchor:arrowView.heightAnchor multiplier:aspectRatio],
      [arrowView.centerXAnchor constraintEqualToAnchor:view.centerXAnchor constant:-60.f],
      [arrowView.centerYAnchor constraintEqualToAnchor:view.centerYAnchor constant:160.f]
    ]];

    label.translatesAutoresizingMaskIntoConstraints = FALSE;
    [NSLayoutConstraint activateConstraints:@[
      [label.topAnchor constraintEqualToAnchor:arrowView.bottomAnchor constant:10.f],
      [label.centerXAnchor constraintEqualToAnchor:view.centerXAnchor],
      [label.leadingAnchor constraintEqualToAnchor:view.leadingAnchor],
      [label.trailingAnchor constraintEqualToAnchor:view.trailingAnchor]
    ]];

    return view;
}

- (void)applyOnboardingViewConstraintsToView:(UIView *)view
                anchorBottomToNextPageButton:(BOOL)anchorToNextPageButton {

    view.translatesAutoresizingMaskIntoConstraints = FALSE;

    NSLayoutConstraint *bottomConstraint;

    if (anchorToNextPageButton) {
        bottomConstraint = [view.bottomAnchor constraintEqualToAnchor:nextPageButton.topAnchor];
    } else {
        bottomConstraint = [view.bottomAnchor constraintEqualToAnchor:progressView.topAnchor
                                                             constant:-20.f];
    }

    [NSLayoutConstraint activateConstraints:@[
      bottomConstraint,
      [view.topAnchor constraintEqualToAnchor:self.view.safeTopAnchor constant:15.f],
      [view.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
      [view.leadingAnchor constraintGreaterThanOrEqualToAnchor:self.view.safeLeadingAnchor],
      [view.trailingAnchor constraintLessThanOrEqualToAnchor:self.view.safeTrailingAnchor],
      [view.widthAnchor constraintLessThanOrEqualToConstant:500.f]  // Max width for large screens.
    ]];
}

@end

File: Psiphon/ViewControllers/OnboardingViewController.swift
/*
 * Copyright (c) 2020, Psiphon Inc.
 * All rights reserved.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import Foundation
import PsiApi
import Utilities
import Promises
import UIKit

/// Enumerates set of all privacy policy versions.
enum PrivacyPolicyVersion: String, Codable {
    
    /// Privacy policy dated 2018-May-15.
    case privacyPolicy_v2018_05_15
    
    /// Privacy policy update on 2021-Oct-06.
    case privacyPolicy_v2021_10_06
    
}

/// Represents different stages of onboarding.
/// Note that each stage may not be limited to only one screen.
enum OnboardingStage: Hashable, RawRepresentable {
    typealias RawValue = String
    
    case languageSelection
    case privacyPolicy(PrivacyPolicyVersion)
    case vpnConfigPermission
    case userNotificationPermission
    
    init?(rawValue: String) {
        
        if let privacyPolicyVersion = PrivacyPolicyVersion(rawValue: rawValue) {
            // rawValue is a privacy policy version
            self = .privacyPolicy(privacyPolicyVersion)
        } else {
            switch rawValue {
            case "languageSelection":
                self = .languageSelection
            case "vpnConfigPermission":
                self = .vpnConfigPermission
            case "userNotificationPermission":
                self = .userNotificationPermission
            default:
                return nil
            }
        }
    }
    
    var rawValue: String {
        switch self {
        case .languageSelection:
            return "languageSelection"
        case .privacyPolicy(let version):
            return version.rawValue
        case .vpnConfigPermission:
            return "vpnConfigPermission"
        case .userNotificationPermission:
            return "userNotificationPermission"
        }
    }
    
    /// Ordered set of the stages that would have to be completed
    /// by the user for this version of the app.
    static let allStages: [OnboardingStage] =
        [ .languageSelection,
          .privacyPolicy(.privacyPolicy_v2021_10_06),
          .vpnConfigPermission,
          .userNotificationPermission ]
    
    /// Returns ordered set of stages not completed by the user, given `completedStages`,
    /// and the current `platform`.
    static func stagesToComplete(
        platform: Platform,
        completedStages: [OnboardingStage]
    ) -> OrderedSet<OnboardingStage> {
        
        var stagesNotCompleted = OnboardingStage.allStages
        
        stagesNotCompleted.removeAll { completedStages.contains($0) }
        
        let hasAcceptedAnyPP = completedStages.contains(where: {
            if case .privacyPolicy(_) = $0 { return true }
            return false
        })
        
        // If the user has already accepted a previous privacy policy,
        // then removes privacy policy stages from `stagesNotCompleted` (if any)
        if hasAcceptedAnyPP {
            stagesNotCompleted.removeAll {
                if case .privacyPolicy(_) = $0 { return true }
                return false
            }
        }
        
        // Removes user notification for Mac.
        stagesNotCompleted.removeAll {
            if case .userNotificationPermission = $0 { return true }
            return false
        }
        
        return OrderedSet(stagesNotCompleted)
    }
    
}

fileprivate extension OnboardingStage {
    
    var screens: OrderedSet<OnboardingScreen> {
        switch self {
        case .languageSelection,
             .privacyPolicy(_):
            return [ OnboardingScreen(stage: self, screenIndex: 0) ]
            
        case .vpnConfigPermission,
             .userNotificationPermission:
            return [ OnboardingScreen(stage: self, screenIndex: 0),
                     OnboardingScreen(stage: self, screenIndex: 1) ]
        }
    }
    
}

// TODO: Temporary struct for when tuples conform to Hashable.
fileprivate struct OnboardingScreen: Hashable {
    let stage: OnboardingStage
    let screenIndex: Int
}

fileprivate extension OnboardingScreen {
    
    var showNextButton: Bool {
        switch (self.stage, self.screenIndex) {
        case (.languageSelection, 0):
            return true
        case (.privacyPolicy(_), 0):
            return false
        case (.vpnConfigPermission, 0):
            return true
        case (.vpnConfigPermission, 1):
            return false
        case (.userNotificationPermission, 0):
            return true
        case (.userNotificationPermission, 1):
            return false
        default:
            fatalError("unknown value '\(self)'")
        }
    }
    
}

@objc final class OnboardingViewController: ReactiveViewController {
    
    /// Ordered set of onboarding stages.
    let onboardingStages: OrderedSet<OnboardingStage>
    
    /// Total number of screens given `onboardingStages`.
    let numberOfScreens: Int
    
    private let userDefaultsConfig: UserDefaultsConfig
    private let mainBundle: Bundle
    
    private let screens: OrderedSet<OnboardingScreen>
    private var currentScreenIndex: Int
    
    /// Convenience property to return current screen with index `currentScreenIndex`.
    fileprivate var currentScreen: OnboardingScreen {
        screens[currentScreenIndex]
    }
        
    // mainLayout defines a virtual layout that contains
    // the main contents of this view controller.
    // For iPhones the mainLayout is the same as self.view layout constraints,
    // for iPads / Macs, the mainLayout is set to constant width and height.
    private let mainLayout = UILayoutGuide()
    
    // Views
    private let versionLabel = UIButton()
    private var currentOnboardingView: UIView? = nil
    private let progressView = UIProgressView(progressViewStyle: .bar)
    private let nextButton = SwiftUIButton(type: .system)
    
    private let platform: Platform
    private let feedbackLogger: FeedbackLogger
    private let installVPNConfig: () -> Promise<VPNConfigInstallResult>
    private let onOnboardingFinished: (OnboardingViewController) -> Void

    init(
        platform: Platform,
        version: String,
        userDefaultsConfig: UserDefaultsConfig,
        mainBundle: Bundle,
        onboardingStages: OrderedSet<OnboardingStage>,
        feedbackLogger: FeedbackLogger,
        installVPNConfig: @escaping () -> Promise<VPNConfigInstallResult>,
        onOnboardingFinished: @escaping (OnboardingViewController) -> Void
    ) {
        guard onboardingStages.count > 0 else {
            fatalError()
        }
        
        self.versionLabel.setTitle(version, for: .normal)
        
        self.userDefaultsConfig = userDefaultsConfig
        self.mainBundle = mainBundle
        
        self.onboardingStages = onboardingStages
        self.numberOfScreens = onboardingStages.map(\.screens.count).reduce(0, +)
        
        self.screens = OrderedSet(onboardingStages.flatMap { $0.screens })
        self.currentScreenIndex = 0
        
        self.platform = platform
        self.feedbackLogger = feedbackLogger
        
        #if targetEnvironment(simulator)
        // Skips VPN config installation since it
        // cannot be installed on a simulator device.
        self.installVPNConfig = {
            Promise { fulfill, _ in
                fulfill(.installedSuccessfully)
            }
        }
        #else
        self.installVPNConfig = installVPNConfig
        #endif
        
        self.onOnboardingFinished = onOnboardingFinished
        
        super.init(onDismissed: {
            // No-op.
        })
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override var preferredStatusBarStyle: UIStatusBarStyle {
        return .lightContent
    }
    
    override var supportedInterfaceOrientations: UIInterfaceOrientationMask {
        return [.portrait, .portraitUpsideDown]
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.setNeedsStatusBarAppearanceUpdate()
        self.view.backgroundColor = .darkBlue()
        
        view.addLayoutGuide(mainLayout)
        
        mainLayout.activateConstraints {
            return $0.constraint(to: self.view.safeAreaAnchors, .centerX(0), .centerY(0)) +
             [
                // mainLayout should always be equal or smaller than safe screen area.
                $0.topAnchor.constraint(greaterThanOrEqualTo: self.view.safeAreaAnchors.topAnchor).priority(.required),
                $0.bottomAnchor.constraint(lessThanOrEqualTo: self.view.safeAreaAnchors.bottomAnchor).priority(.required),
                $0.leadingAnchor.constraint(greaterThanOrEqualTo: self.view.safeAreaAnchors.leadingAnchor).priority(.required),
                $0.trailingAnchor.constraint(lessThanOrEqualTo: self.view.safeAreaAnchors.trailingAnchor).priority(.required),
                
                $0.widthAnchor.constraint(equalToConstant: 468).priority(.belowRequired),
                $0.heightAnchor.constraint(equalToConstant: 736).priority(.belowRequired),
                $0.widthAnchor.constraint(lessThanOrEqualToConstant: 468).priority(.required),
                $0.heightAnchor.constraint(lessThanOrEqualToConstant: 736).priority(.required)
            ]
        }
        
        
        
        // Adds background clouds
        let cloudsView = CloudsView(forAutoLayout: ())!
        self.view.addSubview(cloudsView)
        cloudsView.activateConstraints {
            $0.constraintToParentSafeArea(.top(0), .bottom(0), .leading(0), .trailing(0))
        }
        
        // Adds progress bar to the bottom of the screen
        self.view.addSubview(progressView)
        progressView.progressTintColor = .lightishBlue()
        progressView.activateConstraints {
            $0.constraint(to: mainLayout, .bottom(0), .leading(0), .trailing(0)) +
                [$0.heightAnchor.constraint(equalToConstant: 6.0)]
        }
        
        // Adds next button
        nextButton.setContentEdgeInsets(top: 20, leading: 40, bottom: 20, trailing: 20)
        nextButton.setTitle(Strings.nextPageButtonTitle(), for: .normal)
        nextButton.setTitleColor(.white, for: .normal)
        nextButton.titleLabel!.font = .avenirNextDemiBold(16.0)
        
        self.view.addSubview(nextButton)
        nextButton.activateConstraints {
            $0.constraint(to: mainLayout, .trailing(-20.0, .required)) +
                [ $0.bottomAnchor.constraint(equalTo: progressView.topAnchor) ]
        }
        
        // Versiona label
        let versionLabelPadding: CGFloat = 10.0
        
        mutate(versionLabel) {
            $0.setTitleColor(.nepalGreyBlue(), for: .normal)
            $0.titleLabel?.adjustsFontSizeToFitWidth = true
            $0.titleLabel?.font = .avenirNextBold(12.0)
            $0.isUserInteractionEnabled = false
            $0.contentEdgeInsets = UIEdgeInsets(
                top: versionLabelPadding,
                left: versionLabelPadding,
                bottom: versionLabelPadding,
                right: versionLabelPadding
            )
            #if DEBUG || DEV_RELEASE
            if AppInfo.runningUITest() {
                versionLabel.isHidden = true
            }
            #endif
        }
        
        self.view.addSubview(versionLabel)
        versionLabel.activateConstraints {
            $0.constraint(to: mainLayout,
                .top(Float(Style.default.padding), .required),
                .trailing(Float(-Style.default.padding), .required)
            )
        }
        
        // Displays first screen.
        self.displayScreen(self.currentScreen)
    }
    
    /// Freezes next button, waiting for an async operation to finish.
    private func freezeOnboarding() {
        self.nextButton.isUserInteractionEnabled = false
    }
    
    /// Un-freezes next button so that actions can be taken.
    private func unfreezeOnboarding() {
        self.nextButton.isUserInteractionEnabled = true
    }
    
    private func gotoScreenFollowing(screenIndex: Int) {
        guard self.currentScreenIndex == screenIndex else {
            return
        }
        
        if self.screens[maybe: self.currentScreenIndex + 1] != nil {
            self.currentScreenIndex += 1
            self.displayScreen(self.currentScreen)
        } else {
            // Onboarding finished.
            self.onOnboardingFinished(self)
        }
    }
    
    /// Goes to the starting of the current onboarding stage.
    /// It is an error to call this method if current screen is already the starting screen of current index.
    private func gotoStartOfCurrentStage() {
        guard self.currentScreen.screenIndex > 0 else {
            // Already at the start of the current stage.
            return
        }
        
        self.currentScreenIndex -= 1
        self.displayScreen(self.currentScreen)
    }
    
    private func displayScreen(_ screen: OnboardingScreen) {
        
        let currentIndex = self.currentScreenIndex
        let onboardingView: UIView
        
        switch (screen.stage, screen.screenIndex) {
        case (.languageSelection, 0):
            onboardingView = makeLanguageSelectionOnboardingView { [unowned self] in
                let langSelectionViewController = LanguageSelectionViewController(
                    supportedLocalizations: SupportedLocalizations(
                        userDefaultsConfig: self.userDefaultsConfig, mainBundle: self.mainBundle
                    )
                )
                
                langSelectionViewController.selectionHandler = { _, selectedLang, viewController in
                    
                    guard let selectedLang = selectedLang as? Language else {
                        fatalError()
                    }
                    
                    // Updates user's default app language.
                    userDefaultsConfig.appLanguage = selectedLang.code
                    
                    viewController.dismiss(animated: true, completion: nil)
                    // Reload the onboarding to reflect the newly selected language.
                    AppDelegate.shared().reloadOnboardingViewController()
                }
                
                let nav = UINavigationController(rootViewController: langSelectionViewController)
                self.present(nav, animated: true, completion: nil)
            }

        case (.privacyPolicy(_), 0):
            onboardingView = makeLatestPrivacyPolicyOnboardingView(
                onAccepted: { [unowned self] in
                    self.gotoScreenFollowing(screenIndex: currentIndex)
                },
                onDeclined: { [unowned self] in
                    let alert = makePrivacyPolicyDeclinedAlert()
                    self.present(alert, animated: true, completion: nil)
                }
            )

        case (.vpnConfigPermission, 0):
            
            onboardingView = OnboardingView(
                image: UIImage(named: "OnboardingVPNPermission")!,
                withTitle: Strings.onboardingGettingStartedHeaderText(),
                withBody: Strings.onboardingGettingStartedBodyText(),
                withAccessoryView: nil
            )
            
        case (.vpnConfigPermission, 1):
            onboardingView = makeVPNConfigPermissionGuideOnboardingView()
            
        case (.userNotificationPermission, _):
            
            onboardingView = OnboardingView(
                image: UIImage(named: "OnboardingPushNotificationPermission")!,
                withTitle: UserStrings.Onboarding_user_notification_permission_title(),
                withBody: UserStrings.Onboarding_user_notification_permission_body(),
                withAccessoryView: nil
            )
            
        default:
            fatalError("Unknown onboarding screen '\(screen)'")
        }
        
        self.updateView(onboardingView: onboardingView)
        
        // Updates `nextButton` event handler.
        nextButton.setEventHandler { [unowned self] in
            self.gotoScreenFollowing(screenIndex: currentIndex)
        }
        
        // Carries out any effects based on which screen is presented.
        
        switch screen {
        case OnboardingScreen(stage: .vpnConfigPermission, screenIndex: 1):
            
            self.freezeOnboarding()
            
            self.installVPNConfig().then { [unowned self] vpnConfigInstallResult in
                if Debugging.mainThreadChecks {
                    precondition(Thread.isMainThread, "action not called on main thread")
                }
                
                self.unfreezeOnboarding()
                
                switch vpnConfigInstallResult {
                case .installedSuccessfully:
                    // Go to next onboarding screen.
                    self.gotoScreenFollowing(screenIndex: currentIndex)
                
                case .permissionDenied:
                    // Re-start VPN config permission onboarding stage.
                    self.gotoStartOfCurrentStage()
                    
                    // Presents VPN permission denied alert.
                    let alert = AlertDialogs.vpnPermissionDeniedAlert()
                    self.present(alert, animated: true, completion: nil)
                    
                case .otherError:
                    // Re-start VPN config permission onboarding stage.
                    self.gotoStartOfCurrentStage()
                    
                    // Presents generic operation failed alert
                    let alert = AlertDialogs.genericOperationFailedTryAgain()
                    self.present(alert, animated: true, completion: nil)
                }
            }
            
        case OnboardingScreen(stage: .userNotificationPermission, screenIndex: 1):
            requestUserNotificationPermission(feedbackLogger: self.feedbackLogger) {
                self.gotoScreenFollowing(screenIndex: currentIndex)
            }
            
        default:
            // No effects
            break
        }

    }
    
    private func updateView(onboardingView: UIView) {
        progressView.progress = Float(self.currentScreenIndex + 1) / Float(numberOfScreens)
        
        nextButton.isHidden = !currentScreen.showNextButton
        
        self.currentOnboardingView?.removeFromSuperview()
        self.currentOnboardingView = onboardingView
        self.view.addSubview(onboardingView)
        
        let bottomConstraint: NSLayoutConstraint
        if !nextButton.isHidden {
            bottomConstraint = onboardingView.bottomAnchor
                .constraint(equalTo: self.nextButton.topAnchor)
        } else {
            bottomConstraint = onboardingView.bottomAnchor
                .constraint(equalTo: self.progressView.topAnchor, constant: -20.0)
        }
        
        onboardingView.activateConstraints {
            $0.constraint(to: mainLayout,
                          .top(Float(Style.default.padding)),
                          .centerX(0),
                          .leading(0),
                          .trailing(0)) +
            [ bottomConstraint ]
        }
        
    }
    
}

// MARK: View helper functions

fileprivate func makeLanguageSelectionOnboardingView(
    onLanguageSelected: @escaping () -> Void
) -> OnboardingView {
    let selectLangButton = RingSkyButton(forAutoLayout: ())
    selectLangButton.includeChevron = true
    selectLangButton.setTitle(Strings.onboardingSelectLanguageButtonTitle())
    
    selectLangButton.setEventHandler(onLanguageSelected)
    
    return OnboardingView(
        image: UIImage(named: "OnboardingLanguageSelect")!,
        withTitle: UserStrings.Psiphon(),
        withBody: "",
        withAccessoryView: selectLangButton
    )
}

fileprivate func makeLatestPrivacyPolicyOnboardingView(
    onAccepted: @escaping () -> Void,
    onDeclined: @escaping () -> Void
) -> OnboardingScrollableView {
    let acceptButton = RoyalSkyButton(forAutoLayout: ())
    acceptButton.setTitle(UserStrings.Accept_button_title())
    acceptButton.shadow = true
    acceptButton.setEventHandler(onAccepted)
    
    let declineButton = RingSkyButton(forAutoLayout: ())
    declineButton.setTitle(UserStrings.Decline_button_title())
    declineButton.setEventHandler(onDeclined)
    
    let stackView = UIStackView(arrangedSubviews: [declineButton, acceptButton])
    stackView.spacing = 20.0
    stackView.distribution = .fillEqually
    
    return OnboardingScrollableView(
        image: UIImage(named: "OnboardingPrivacyPolicy")!,
        withTitle: Strings.privacyPolicyTitle(),
        withHTMLBody: UserStrings.privacyPolicyHTMLText_v2021_10_06(languageCode: UserDefaultsConfig().localeForAppLanguage.languageCode ?? ""),
        withAccessoryView: stackView
    )
}

fileprivate func makePrivacyPolicyDeclinedAlert() -> UIAlertController {
    let alertController = UIAlertController(
        title: Strings.privacyPolicyTitle(),
        message: Strings.privacyPolicyDeclinedAlertBody(),
        preferredStyle: .alert
    )
    alertController.addAction(
        UIAlertAction(title: UserStrings.Dismiss_button_title(), style: .cancel)
    )
    return alertController
}

/// Creates view for display under system VPN config permission dialog.
/// The X and Y centre of the returned view is expected to match the X and Y centre of the screen.
fileprivate func makeVPNConfigPermissionGuideOnboardingView() -> UIView {
    let view = UIView()
    let label = UILabel.make(text: Strings.vpnInstallGuideText(),
                             fontSize: .h3,
                             typeface: .medium,
                             color: .white,
                             numberOfLines: 0,
                             alignment: .center)
    view.addSubviews(label)
    label.activateConstraints {
        $0.constraint(to: view, .centerX(0), .bottom(-100), .leading(0), .trailing(0))
        
    }
    return view
}

fileprivate func requestUserNotificationPermission(
    feedbackLogger: FeedbackLogger,
    completionHandler: @escaping () -> Void
) {
    let centre = UNUserNotificationCenter.current()
    centre.getNotificationSettings { settings in
        guard settings.authorizationStatus == .notDetermined else {
            DispatchQueue.main.async {
                completionHandler()
            }
            return
        }
        
        centre.requestAuthorization(options: [.alert, .badge]) { granted, maybeError in
            feedbackLogger.immediate(
                .info, "UserNotification authorization granted: \(granted)")
            
            if let error = maybeError {
                feedbackLogger.immediate(
                    .error, "user notification authorization error: '\(error)'")
            }
            
            DispatchQueue.main.async {
                completionHandler()
            }
        }
    }
}
